hal/src/common/dmac/channel.rs

```diff
@@ -0,0 +1,313 @@
+//! # Abstractions over individual DMA channels
+//!
+//! # Initializing
+//!
+//! Individual channels should be initialized through the
+//! [`Channel::init`](Channel::init) method. This will return a `Channel<Ready,
+//! ID>` ready for use by a [`DmaTransfer`](super::transfer::DmaTransfer).
+//! Initializing a channel requires setting a priority level, as well as
+//! enabling or disabling interrupt requests (only for the specific channel
+//! being initialized).
+//!
+//! # Burst Length and FIFO Threshold (SAMD51/SAME5x only)
+//!
+//! The transfer burst length can be configured through the
+//! [`burst_length`](Channel::burst_length) method. A burst is an atomic,
+//! uninterruptible transfer which length corresponds to a number of beats. See
+//! SAMD5x/E5x datasheet section 22.6.1.1 for more information. The FIFO
+//! threshold can be configured through the [`fifo_threshold`](Channel::
+//! fifo_threshold) method. This enables the channel to wait for multiple
+//! Beats before sending a Burst. See SAMD5x/E5x datasheet section 22.6.2.8 for
+//! more information.
+//!
+//! # Channel status
+//!
+//! Channels can be in any of three statuses: `Uninitialized`, `Ready`, and
+//! `Busy`. These statuses are checked at compile time to ensure they are
+//! properly initialized before launching DMA transfers.
+//!
+//! # Resetting
+//!
+//! Calling the [`reset`](Channel::reset) method will reset the channel to its
+//! `Uninitialized` state. You will be required to call [`init`](Channel::init)
+//! again before being able to use it with a `DmaTransfer`.
+
+use super::dma_controller::{DmaController, PriorityLevel, TriggerAction, TriggerSource};
+
+#[cfg(any(
+    feature = "samd51",
+    feature = "same51",
+    feature = "same53",
+    feature = "same54"
+))]
+use super::dma_controller::{BurstLength, FifoThreshold};
+
+#[cfg(any(
+    feature = "samd51",
+    feature = "same51",
+    feature = "same53",
+    feature = "same54"
+))]
+use crate::target_device::dmac::CHANNEL;
+
+use crate::target_device::DMAC;
+use crate::typelevel::Sealed;
+
+/// Uninitialized channel
+pub struct Uninitialized;
+/// Initialized and ready to transfer channel
+pub struct Ready;
+/// Busy channel
+pub struct Busy;
+
+/// DMA channel, capable of executing
+/// [`DmaTransfer`](super::transfer::DmaTransfer)s.
+pub struct Channel<Status, const ID: u8> {
+    _status: Status,
```

Comment:
This isn't strictly necessary, but I think it might help clarify intent and help prevent mistakes when refactoring. I would create a `Status` trait (with `Sealed` as a super trait). Then, I would implement `Status` for `Uninitialized`, `Ready` and `Busy`. Finally, I would have `Channel` accept a type parameter `S`, `where S: Status`.

Effectively, the status trait is like a meta-enum, and the `Uninitialized`, `Ready` and `Busy` types are variants of that type-level enum. To me, that makes the intent really clear.